File: lxc_exporter.py
#!/usr/bin/python

import subprocess
from flask import Flask
app = Flask(__name__)

def run(command):
    p = subprocess.Popen(command, stdout=subprocess.PIPE, shell=True)
    return p.communicate()[0]

def containers():
    containers = []
    for lxc in run("lxc-ls --active -1").split("\n"):
        if lxc:
            containers.append(int(lxc.strip()))
    return containers

def stats(container):
    stats = {}
    for lxc in run("lxc-info -S -H -n %d" % container).split("\n"):
        if lxc:
            key, value = lxc.split(":")
            value = value.strip()
            stats[key.strip()] = int(value) if value.isdigit() else value
    return stats

def ips(container):
    ips = []
    for lxc in run("lxc-info -i -n %d" % container).split("\n"):
        if lxc:
            _, value = lxc.split("IP:")
            ips.append(value.strip())
    return ips

def info():
    info = []
    metrics = {
        'cpu': 'CPU use',
        'memory': 'Memory use',
        'total_bytes': 'Total bytes',
        'rx_bytes': 'RX bytes',
        'tx_bytes': 'TX bytes',
        'io': 'BlkIO use'
    }
    for lxc in containers():
       for metric, value in metrics.iteritems():
           info.append(to_prometheus(metric, lxc, ips(lxc)[1], stats(lxc)[value]))
    return info

def to_prometheus(metric, id, ip, value):
    return 'node_lxc_%s{id="%s", ip="%s"} %s\n' % (metric, id, ip, value)

@app.route("/")
def root():
    return "<h1>LXC metrics</h1><br><a href=/metrics>Metrics</a>"

@app.route("/metrics")
def metrics():
    return "".join(info()), 200, {'Content-type': 'text/plain'}

if __name__ == "__main__":
    app.run(host='::', port=9119)

File: lxc_exporter.go
package main

import (
    "flag"
    "net/http"
    "sync"
    "os/exec"
    "strconv"
    "strings"
    "github.com/prometheus/client_golang/prometheus"
    "github.com/prometheus/common/log"
)

const (
    namespace = "lxc"
)

var (
    listeningAddress    = flag.String("telemetry.address", ":9119", "Address on which to expose metrics")
    metricsEndpoint     = flag.String("telemetry.endpoint", "/metrics", "Path under which to expose metrics")
    containerLabels     = []string{"id"}
    metrics             = map[string]*prometheus.Desc {
        "cpu":          newDesc("cpu", containerLabels),
        "memory":       newDesc("memory", containerLabels),
        "total_bytes":  newDesc("total_bytes", containerLabels),
        "rx_bytes":     newDesc("rx_bytes", containerLabels),
        "tx_bytes":     newDesc("tx_bytes", containerLabels),
        "io":           newDesc("io", containerLabels),
    }
    metricsMapping      = map[string]string {
        "cpu":          "CPU use",
        "memory":       "Memory usage",
        "total_bytes":  "Total bytes",
        "rx_bytes":     "RX bytes",
        "tx_bytes":     "TX bytes",
        "io":           "BlkIO use",
    }
)

type Exporter struct {
    mutex sync.Mutex
    up *prometheus.Desc
    metricsCounter map[string]*prometheus.Desc
    scrapeFailures prometheus.Counter
}

func newDesc(metricName string, labels []string) *prometheus.Desc {
	return prometheus.NewDesc(
        prometheus.BuildFQName(namespace, "", metricName),
        metricsMapping[metricName],
        labels,
        nil)
}

func Containers() []int {
    var containers = []int{}
    output, err := exec.Command("lxc-ls", "--active", "-1").Output()

    if err != nil {
        log.Errorf("Error while listing LXC containers: ", err)
    }

    for _, id := range strings.Split(string(output), "\n") {
        if id != "" {
            i, err := strconv.Atoi(id)
            if err != nil {
                log.Errorf("Cannot parse LXC id: ", err)
            }
            containers = append(containers, i)
        }
    }
    return containers
}

func NewExporter() *Exporter {
    return &Exporter{
        up: prometheus.NewDesc(
            prometheus.BuildFQName(namespace, "", "up"),
            "Could the LXC exporter be reached",
            nil,
            nil),
        scrapeFailures: prometheus.NewCounter(prometheus.CounterOpts{
            Namespace: namespace,
            Name:      "exporter_scrape_failures_total",
            Help:      "Number of errors while scraping LXC containers",
            }),
        metricsCounter: metrics,
    }
}

func (e *Exporter) Describe(ch chan<- *prometheus.Desc) {
        ch <- e.up
        e.scrapeFailures.Describe(ch)
    for _, metric := range e.metricsCounter {
        ch <- metric
    }
}

func (e *Exporter) collect(ch chan<- prometheus.Metric) error {
    ch <- prometheus.MustNewConstMetric(e.up, prometheus.GaugeValue, 1)
    for _, id := range Containers() {
        container_id := strconv.Itoa(id)
        output, err := exec.Command("lxc-info", "-S", "-H", "-n", container_id).Output()

        if err != nil {
            log.Errorf("Error while getting stats for container (%d): ", container_id, err)
        }

        for _, info := range strings.Split(string(output), "\n") {
            if info != "" {
                info := strings.Split(info, ":")
                metric := strings.Trim(info[0], " ")
                valueString := strings.Trim(info[1], " ")
                valueInt, _ := strconv.ParseFloat(valueString, 64)
                for key, val := range metricsMapping {
                    if metric == val {
                        ch <- prometheus.MustNewConstMetric(e.metricsCounter[key], prometheus.CounterValue, valueInt, container_id)
                    }
                }
            }
        }
    }
    return nil
}

func (e *Exporter) Collect(ch chan<- prometheus.Metric) {
    e.mutex.Lock()
    defer e.mutex.Unlock()
    if err := e.collect(ch); err != nil {
        log.Errorf("Error scraping: %s", err)
        e.scrapeFailures.Inc()
        e.scrapeFailures.Collect(ch)
    }
    return
}

func main() {
    flag.Parse()

    exporter := NewExporter()
    prometheus.MustRegister(exporter)

    log.Infof("Starting Server: %s", *listeningAddress)
    http.Handle(*metricsEndpoint, prometheus.Handler())
    log.Fatal(http.ListenAndServe(*listeningAddress, nil))
}
